CASSANDRA-14459: DynamicEndpointSnitch should never prefer latent replicas

jolynch opened the pull request:
This change incorporates the feedback from Ariel and Jason as part of https://issues.apache.org/jira/browse/CASSANDRA-14459.

The following is introduced:

1. Fully pluggable DynamicEndpointSnitch so that we can continue experimenting with new implementations
2. Instead of resetting every 10 minutes, the DES uses active latency probes for replicas that it was asked to rank but has no recent data on. These are rate limited by default to a single probe per second. These latency probes, while not perfect, will correctly detect nodes that are latent due to network conditions, JVM instability (gc/safepoint pauses), and Read threadpool exhaustion.
3. A new opt-in implementation of the DES which uses an exponential moving average instead of a Histogram. Both statistical measures try to develop a noise reduced sample with different tradeoffs, but the main one in favor of DES is that it reacts to extreme outliars faster (e.g. if a node is actively timing out and dropping messages) and generates about 100x less garbage than the histogram approach.

jolynch committed:
Avoid dropping all data in DynamicSnitch reset

Instead of throwing away all measurements every ten minutes, now we keep
the old value and proactively probe nodes that are being requested
for ranking but are not getting any latency information using
exponential backoff. We use these probes to re-incorporate slow nodes
rather than a blind reset

This prevents nodes from talking accross datacenters and makes it so that when
nodes start up they get a latency landscape during the first round of startup
ping messages.

aweisberg commented on doc/source/operating/snitch.rst:
```diff
@@ -28,16 +28,18 @@ In cassandra, the snitch has two functions:
-- ``dynamic_snitch_reset_interval_in_ms``: if set greater than zero, this will allow 'pinning' of replicas to hosts
-  in order to increase cache capacity.
-- ``dynamic_snitch_badness_threshold:``: The badness threshold will control how much worse the pinned host has to be
+- ``dynamic_snitch_latency_probe_interval_in_ms``: controls how often this node
```
Document dynamic snitch class name?

jolynch replied:
Hm, is this pluggability something that we want to expose to users? I thought that it was just going to be so that devs can experiment with different implementations without going through a cassandra release every time (e.g. we can actually run various versions "in production").

If you think it's a good idea to let users know that they can plugin their own DES class, I can make the change, but I thought it was just a way to allow us to experiment until we actually have good data on which approach is "best".

aweisberg replied:
OK. I agree we don't want to encourage people to implement their own dynamic snitch.

aweisberg commented on conf/cassandra.yaml:
```diff
@@ -947,10 +947,10 @@ endpoint_snitch: SimpleSnitch
+dynamic_snitch_update_interval_in_ms: 100
+# controls how often this node is permitted to send a latency probe to
+# determine if previously slow nodes have recovered. Set to -1 to disable
+dynamic_snitch_latency_probe_interval_in_ms: 1000
```
Add dynamic snitch class name as a commented out option?

jolynch replied:
I'm resolving this since I think we agreed not to do this.

aweisberg commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -224,8 +224,11 @@
     public boolean dynamic_snitch = true;
+    public String dynamic_snitch_class_name = "org.apache.cassandra.locator.DynamicEndpointSnitchHistogram";
     public int dynamic_snitch_update_interval_in_ms = 100;
+    @Deprecated
```
If it's unused you can change it to an Integer and set it to null and then if it's non-null log a warning. You will also need to update NEWS.txt to document what happened with the snitch configuration and active probing. I would wait until it's closer to done to add it.

jolynch replied:
Ok, I've done the first part (making it an Integer and logging a warning). I will do the NEWS.txt update at the end like you recommend.

aweisberg commented on src/java/org/apache/cassandra/net/IAsyncCallback.java:
```diff
@@ -31,11 +31,23 @@
      * @return true if this callback is on the read path and its latency should be
      * given as input to the dynamic snitch.
      */
+    @Deprecated
```
You don't need to deprecate you can just remove it? It's purely internal?

jolynch replied:
Removing it will cause a lot of changes, I can do that but it will make the diff much larger (I see about 20 files that would need changes)

aweisberg replied:
Yes we have way too many mock snitches. I think there are some things that require one to be constructed. I would still make the change, and maybe if possible consolidate the mock snitches into as few as possible? We definitely don't want to keep around deprecated internal code and we have to pay whatever the cost is to get rid of it.

jolynch replied:
Ok, I've done the refactor. I'd like to keep it as a separate commit though so that backporting is easier if that's ok?

aweisberg replied:
Sure I'll squash before merging to trunk

aweisberg commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -38,6 +38,7 @@
 import com.google.common.util.concurrent.Uninterruptibles;
 
 import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.net.LatencyMeasurementType;
```
Unused import

jolynch replied:
ack, fixed

aweisberg commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -1055,7 +1056,6 @@ public boolean isLatencyForSnitch()
             {
                 return false;
             }
-
```
This whitespace change doesn't make sense?

jolynch replied:
ack, got lost in the various additions and removals of whether EchoMessages counted or not

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -22,121 +22,194 @@
+    protected volatile Map<InetAddressAndPort, Double> scores = new HashMap<>();
+
+    // Rate limit how often we generate latency probes
+    protected long nextAllowedProbeGenerationTime;
```
As we discussed I don't think we really need this. We can back off probing nodes we aren't talking to very frequently to get back to the only check a node every 10 minutes behavior. That coupled with not probing nodes we aren't even considering for queries should do get the messaging down.

jolynch replied:
I dig backoff, implementing.

jolynch replied:
Alright, exponential backoff implemented and indeed it reduces the messaging significantly. The current patch only starts probing nodes that have been requested in the last ten minutes, but have not had latency measurements in the last one minute. These probes exponentially backoff until ten minutes and then they either keep going at that pace if the host is still being requested, or stop if the host is no longer being requested.

I've left a bunch of comments in the code to explain the algo if it's unclear.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -22,121 +22,194 @@
+     * on hosts that this peer would potentially talk to.
+     * @param address
+     */
+    abstract protected void markRequested(InetAddressAndPort address);
```
I don't get why mark requested is in the derived class? Seems like this is a property of the probing approach that is common to the different implementations.

jolynch replied:
It was there because I had the wrapped metrics (reservoir/EMA) in the subclasses (as they don't both e.g. conform to some interface that I can then compose on top of). I'm trying to implement your suggestion of moving samples back to base class and along the way I wouldn't need this anymore.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -22,121 +22,194 @@
+     * that will be assigned to the {@link DynamicEndpointSnitch#scores} map.
+     * This is generally more expensive than {@link DynamicEndpointSnitch#receiveTiming}
+     */
+    abstract public Map<InetAddressAndPort, Double> calculateScores();
```
This calculation is the same as well across implementations? Or it should be? The only difference is what we are using to store timings and what it spits out?

jolynch replied:
I was thinking that it would be pretty different in future implementations, but I can just make it protected and then subclasses can redefine it anyways :-)

aweisberg replied:
So I think future implementations just override the superclass method if they want?

aweisberg replied:
Additionally, the might not end up being the super class for future snitch implementations since they can implement IEndpointSnitch and replace whatever they want. All we have to worry about today is solving for EMA and Histogram I think.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -22,121 +22,194 @@
+    /**
+     * Records a latency. This MUST be cheap as it is called in the fast path
+     */
+    abstract public void receiveTiming(InetAddressAndPort address, long latency, LatencyMeasurementType measurementType);
```
This should also be common across derived classes except for the how it stores the latency value in the end?

jolynch replied:
correct, and storing the latency is basically all either implementation does. As part of pulling samples (along with the annotations) back to the base class I'll see how much of this I can unify.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -22,121 +22,194 @@
+    /**
+     * Populates the provided probe sequence using the underlying metrics
+     */
+    abstract void updateLatencyProbeSequence(List<InetAddressAndPort> sequence);
```
Again this seems like a super class concern.

jolynch replied:
Agreed assuming I can pull the samples back to the superclass, working on it

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,20 +228,101 @@ private void registerMBean()
+    private void sendPingMessageToPeer(InetAddressAndPort to)
+    {
+        logger.trace("Sending a small and large PingMessage to {}", to);
+        IAsyncCallback latencyProbeHandler = new IAsyncCallback()
```
You can create two settable futures here (or the completable equivalent) and separate callbacks for small and large that point to the settable futures. You can then create a composite future and add a listener to it that does the right thing with the resulting largest/worst latency value.

I think this means you still need the enum, but we don't actually need to change the API of callbacks since we will never report a probe via a callback directly.

jolynch replied:
Hm, ok I'm not entirely sure I follow but I will take a whack at it and get back to you!

jolynch replied:
In the latest version I did this max math but kept the changed callback API but I can remove it you think it'll make the PR easier to review/merge.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -188,6 +343,9 @@ public String getDatacenter(InetAddressAndPort endpoint)
     public <C extends ReplicaCollection<? extends C>> C sortedByProximity(final InetAddressAndPort address, C unsortedAddresses)
     {
         assert address.equals(FBUtilities.getBroadcastAddressAndPort()); // we only know about ourself
+        for (Replica unsortedAddress : unsortedAddresses)
```
We shouldn't need to fetch each endpoint from the map a second time.

jolynch replied:
So I can easily eliminate the extra iteration when calling into `sortedByProximityWithBadness`, but `sortedByProximityWithScore` only calls `compareEndpoints` and putting two more dict lookups there seemed likely slower (to me) than just iterating the unsortedAddresses in `sortedByProximity`. I can make this method more complicated to special case `sortedByProximity -> sortedByProximityWithScore` (as opposed to the normal path of `sortedByProximity` -> sortedByProximityWithBadness -> sortedByProximityWithScore`) ... sure

aweisberg replied:
Hmm... yeah... seems like the score should maybe be merged with the other map containing the reservoir and the score is just an attribute of the annotated thing.

jolynch replied:
I don't think that combining the two maps is a good idea and this loop seems pretty cheap to me whether we do it here first or add a statement to a later loop that reads the score map. Due to the way that we alias scores to and do full copy on write I'm also worried about correctness given the number of comments explaining why that scores aliasing is so important I don't want to risk changing that in this patch.

I do agree a through examination of the scores datastructure and way that we update it may be warranted in the future (e.g. we could create some kind of immutable view just for the endpoints we're ranking, although I don't know a way to do that without adding lots of extra garbage).

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -210,8 +368,10 @@ public String getDatacenter(InetAddressAndPort endpoint)
-                                                           // (which wouldn't really matter here but its cleaner that way).
+        // Make sure the score don't change in the middle of the loop below
+        // (which wouldn't really matter here but its cleaner that way).
+        final Map<InetAddressAndPort, Double> scores = this.scores;
```
I don't think we generally do final on the stack or for parameters to functions. I can't find anything in the style guide that says you can't do it, but I recall being told that at one point. I'll ask around.

jolynch replied:
Yea I was somewhat surprised to see that too, I assumed there was a reason that it was final in `sortedByProximityWithBadness` and just made `sortedByProximityWithScores` consistent, but tbh I don't care and will switch to whichever way is better for you ;-)

aweisberg replied:
I asked and yes we do strongly discourage final on the stack and on method parameters.

jolynch replied:
Cool, I made them consistent the other way.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitchEMA.java:
```diff
@@ -0,0 +1,181 @@
+     *
+     * Both markers are periodically reset via {@link DynamicEndpointSnitch#latencyProbeNeeded(long)}
+     */
+    private static class AnnotatedEMA extends ExponentialMovingAverage
```
You can use a common base with and compose instead of extending so that the base class can contain all the type specific functionality. I think you want to wrap anyways so we can adapt the methods from the underlying reservoir.

Then the map can be in the base class since it's basically common anyways.

jolynch replied:
Ok, makes sense, refactoring :-)

jolynch replied:
This should be done. Almost everything is in the super classes (only method is one that provides an instance of `ISnitchMeasurement` for the superclass).

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitchEMA.java:
```diff
@@ -0,0 +1,181 @@
+
+        public void update(long value, boolean isRealRead)
+        {
+            recentlyMeasured = recentlyMeasured || isRealRead;
```
This can be |=. This isn't checking of recently measured is already set before writing which means it's always going to pull the cache line in for write.

isRealRead sounds a little awkward and not proscriptive. What we are really trying to say is that it should count as recently measured. I think it might be clearer if it simply didn't override update.

I am also a developing fan of enums instead of booleans since true/false is just not that informative.

jolynch replied:
Ah, I interpreted "store extra boolean" flags as meaning you wanted booleans, but I guess what you mean is C style enum flags? I dig it.

Changing the style to Tequals as now we're doing bitwise flags.

aweisberg replied:
You are right I did say boolean. The enum thing occurred to me later as it's usually a response to seeing a boolean in a method call and being "what does that do?" It's a new habit I am trying to build.

jolynch replied:
I ended up making these two `AtomicIntegers` to make the state really clear and the code easier to understand per your recommendations to use a STPE instead of doing it all on the one thread.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitchEMA.java:
```diff
@@ -0,0 +1,181 @@
+    {
+        AnnotatedEMA ema = samples.get(address);
+        if (ema != null)
+            ema.recentlyRequested = true;
```
When we set this boolean we also want to check if it's already set before setting it.

jolynch replied:
I was curious about this if a load before the set would be faster or slower than just a set. I guess I decided that the performance difference was probably so negligible I should just go for most readable. I don't care though heh, changing with the flag update.

aweisberg replied:
It's about scalability more than performance. Every time you write to the cache line it has to be acquired in an exclusive state to write and it's invalidated in caches everywhere else. For volatile fields that are set to the same value repeatedly at high frequency it can be better to not do that.

It's more about avoiding death by a thousand cuts WRT to performance. If it were really really not clear I would say don't do it, but it's a simple enough pattern for setting a boolean.

jolynch replied:
With the latest change I do a `get` before I do a `lazySet` on the timekeeping state. This way within a given second we only do a single `lazySet` operation (allowing gets to hit the cache basically always).

I also did a quick JMH run to confirm that the load barrier is cheaper than the unconditional store-store barrier (`lazySet`). It is indeed better by a decent amount.

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -665,9 +665,9 @@ public void updateBackPressureOnSend(InetAddressAndPort host, IAsyncCallback cal
      * Updates the back-pressure state on reception from the given host if enabled and the given message callback supports it.
      *
-     * @param host The replica host the back-pressure state refers to.
+     * @param host     The replica host the back-pressure state refers to.
```
Your IDE is set up to fix whitespace on unchanged lines. That's generally not great just because it introduces a lot of extra changes, and if two people have IDEs with different preferences they end up fighting and introducing random conflicts all the time.

Can you go through this change set and remove whitespace changes for things that haven't actually been touched?

jolynch replied:
Interesting, I'm just using the default style guide for idea that came with Cassandra (`ant generate-idea-files`). If it's just that review is tough you can tell github not to show you whitespace changes.

That's good feedback though that you don't want unrelated whitespace changes, I'll take it out :-)

aweisberg replied:
If you go to IntelliJ preferences, search for saved, under Editor -> General -> Strip trailing spaces you can select modified lines from the drop down. Not sure if that impacts this particular reformatting here.

jolynch replied:
Ack, all setup thanks :-)

aweisberg commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -469,15 +469,22 @@
+     * @param dynamicUpdateInterval      integer, in ms (defaults to the value configured in cassandra.yaml, which defaults to 100)
+     * @param dynamicLatencyProbeInterval integer, in ms (defaults to the value configured in cassandra.yaml, which defaults to 1,000)
      */
+    public void updateSnitch(String epSnitchClassName, Double dynamicBadnessThreshold, Boolean dynamic, Integer dynamicUpdateInterval, Integer dynamicLatencyProbeInterval) throws ClassNotFoundException;
```
They can specify a snitch class name, but not a dynamic snitch class name?
In general they can't specify much in terms of arbitrary config for the DES they custom specified earlier. Seems like pluggability without configurability seems to be only half way there. Seems like they should be able to provide a map containing the config.

I mean maybe custom snitches are configured some other way, but that turns updating the snitch via JMX into a two step process. Once to satisfy C* and again to satisfy your custom snitch implementation.

jolynch replied:
Hm, like I said before, I didn't want to expose the DES pluggability in the JMX interface since I wasn't sure that we wanted to support it as a user interface (I just want to give a path forward for devs to experiment with different implementations which e.g. use different measurement techniques or load (un)balancing techniques).

Like you say, `IEndpointSnitch`'s are not configurable currently (they have to have an empty constructor that e.g. calls their own `loadConfiguration` method a la `GossipingPropertyFileSnitch`. I think requiring custom DES implementations to do something similar isn't unreasonable (and it reduces complexity).

aweisberg replied:
OK that sounds reasonable.

aweisberg reviewed (changes requested):
I think there is a fair amount of code in the derived classes that could be moved to the base class? Am I wrong here? They should only differ in a small number of ways.

I edited this comment because I accidentally posted an earlier draft that was way off base.

jolynch committed:
Remove deprecated isLatencyForSnitch and replace with latencyMeasurementType

aweisberg commented on src/java/org/apache/cassandra/locator/dynamicsnitch/DynamicEndpointSnitchEMA.java:
```diff
@@ -0,0 +1,133 @@
+        // We're going to weight the latency for each host against the worst one we see, to
+        // arrive at sort of a 'badness percentage' for them. First, find the worst for each:
+        HashMap<InetAddressAndPort, Double> newScores = new HashMap<>(samples.size());
+        Optional<Double> maxObservedAvgLatency = samples.values().stream()
```
This iterates the measurements twice while they are changing and that can result in the maximum no longer being correct so they won't be scaled to 1.0. This means that severity is no longer relative to 1.0.

Not a huge deal but it's also a lot of code duplication to avoid allocating the extra map.

jolynch replied:
I went ahead and removed the EMA entirely, let's follow up on that in another change.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,20 +310,155 @@ private void registerMBean()
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            AnnotatedMeasurement measurement = entry.getValue();
+            long intervalsSinceLastMeasure = measurement.intervalsSinceLastMeasure.getAndIncrement();
```
I am so confused now. This doesn't work because it only sends one probe every 10 seconds. To probe every node in a 1k node non-token aware cluster it would take 1000 * 10 / 60 = 166 minutes. Do we need to do the every ten seconds thing if we are also implementing exponential backoff? Once it's backed off it's 1.6 probes/second which seems totally reasonable.

Also I don't really understand why we are doing this in terms of measuring intervals when what we are trying to track here is time. Every time we send a probe we can double the time to the next probe, and calculate the time in the future when the next probe should occur. 

This seems a more complicated than it has to be. A single threaded scheduled threadpool executor with a RateLimiter seems like it could do this much more clearly.

Receive timing could check for a ScheduleFuture, if it exists cancel it and null it out. The scheduled future can reschedule itself if no real timings have been received.

The existing STPE can check at the minimum interval if any nodes need to start being probed and create the scheduled future. It can also do the cleanup for live nodes.

Determining constant vs exponential can be done by applying Math.min to the result of doubling the backoff vs the limit.

Another plus is because everything is probed independently you don't need to shuffle anymore. We can set the rate limit to place a ceiling on how much intra-cluster probing can occur if there is a thundering herd. Like maybe 5 per second?

jolynch replied:
I believe that I've addressed this feedback, although I didn't end up looking at the futures in receiveTiming since that's performance critical (we'll maybe send an extra probe, not a big deal imo)

jolynch committed:
Feedback from Review

Makes all the bookeeping into time accounting, adds a ton of unit tests
and fixes the Legacy behavior to actually be the legacy behavior.

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4940,42 +4940,61 @@ public int getDynamicUpdateInterval()
-            DatabaseDescriptor.setDynamicResetInterval(dynamicResetInterval);
         if (dynamicBadnessThreshold != null)
             DatabaseDescriptor.setDynamicBadnessThreshold(dynamicBadnessThreshold);
+        if (dynamicSampleUpdateInterval != null)
```
Wow I wasn't expecting you to make it a hot prop. I thought we had left it that it was fine without it. Good to have it since people can go back to the old snitch without bouncing the cluster.

jolynch replied:
That's what I figured, also makes testing a lot easier. I added a bunch of tests of the various swaps as well.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -19,34 +19,104 @@
         }
-        Thread.sleep(150);
+        // Slightly higher than the update interval to allow scores to propagate
+        Thread.sleep(UPDATE_INTERVAL_MS + 10);
```
This could end up being a source of flakiness. A test method to force it to recalculate in the other thread and wait on the resulting future.

jolynch replied:
Yea I mean the status quo was this could be pretty flakey, one of the reasons that I separated out `updateScores` was so that we can forcibly recalculate scores using it. I went ahead and switched this method to just forcibly call `updateScores`, I guess we don't need a test that the threading works?

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
-                updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
+                if (updateScoresScheduler != null)
+                    updateScoresScheduler.cancel(false);
+                updateScoresScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateScores, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
```
Wow so we do this on the scheduled tasks threadpool? STPE doesn't add threads and it's a core pool size is 1. I wonder if other things get flakey because of that. There are quite a few things using it!

aweisberg replied:
Should this be more "maybeScheduleProbes"

jolynch replied:
Yea ... I can move it onto a dedicated STPE but I didn't want to change too much behavior. Regarding the naming the LegacyHistogram doesn't send any probes, so I thought it could just be generically a `updateSamples` method similar to `updateScores` ... I don't care strongly one way or the other it's up to you :-)

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
+                if (newDynamicLatencyProbeInterval > 0)
+                {
+                    probeRateLimiter.setRate(dynamicLatencyProbeInterval);
+                    updateSamplesScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateSamples, dynamicLatencyProbeInterval, dynamicLatencyProbeInterval, TimeUnit.MILLISECONDS);
```
Now that we have this a thread for pings we can also do the scheduling work on it.

jolynch replied:
I considered that but didn't like it because then we can't block in the acquire of the ratelimiter tokens (as that would block the logical clock increments). I think that the amount of work we're doing here is on the same order as before (iterating the samples map) so we should be ok right?

aweisberg replied:
Hmm... really I just wanted to not be blocked behind all the other things in that other thread pool. Where is an asynchronous rate limit when you need it. I suppose it's fine since we only need it to run on the order of seconds and we can stand to be delayed a little.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
-        };
-        reset = new Runnable()
+            applyConfigChanges(DatabaseDescriptor.getDynamicUpdateInterval(),
+                               DatabaseDescriptor.getDynamicSampleUpdateInterval(),
```
This config option doesn't look like it hooks up to the rate limiter the way you intended?

jolynch replied:
It's set within `applyConfigChanges`, I'll reorganize that method to make it more readable.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+            long delay = measurement.probeTimerMillis;
+            long millisSinceLastRequest = measurement.millisSinceLastRequest.get();
+
+            if (millisSinceLastRequest > MAX_PROBE_INTERVAL_MS && !Gossiper.instance.isAlive(entry.getKey()))
```
Nit but this feels like it should be in calculateProbes

jolynch replied:
This way all of the global state (`gossiper`) and threading state (`latencyProbeExecutor`) are in this method instead of `calculateProbes`, which means that we can call `calculateProbes` without checking if the gossiper is active (to maintain the logical clock moving forward).

If we want to calculate the elapse time as you suggest instead of relying on the scheduler to schedule us ~1s then this becomes less of a problem.

aweisberg replied:
I think I like the simplicity of where things are right now and I want to take it back. It's still mostly bounded on how far off it can be from the configured value and it won't go too fast. That should be fine?

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+    {
+        probeRateLimiter.acquire(dynamicLatencyProbeInterval);
+
+        if (!StorageService.instance.isGossipActive())
```
Just curious why this check? Maybe acquire the rate limiter after this check?

jolynch replied:
Because this thread can run up to 10 minutes after it was scheduled, we maybe shouldn't be talking to other nodes at this junction (i.e. gossip is turned off). I know it's racey in the sense that we can check, then gossip can be turned off, then we can send a ping; but realistically users are going to be having issues where they're trying to drain the node and then still get messages 10 minutes later out of this.

Ack on moving it above the rate limit acquire, I moved it before the acquire just to let other probes go ahead.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+                continue;
+
+            AnnotatedMeasurement measurement = entry.getValue();
+            long lastMeasure = measurement.millisSinceLastMeasure.getAndAdd(intervalMillis);
```
Should these be addAndGet?

jolynch replied:
I don't care too much now that `MIN_PROBE_INTERVAL` is 60s, when it's smaller (like 1s) I felt that we were probing too soon. I'll change it.

aweisberg replied:
Hmm that got me thinking, wouldn't a useful JMX gauge be probes sent per second? I am wondering if that is a nice to have just so you can monitor how much work is being done by the snitch? WDYT?

aweisberg replied:
Sorry that is a silly request. We hard rate limit it so there isn't much of a reason to monitor for how often we do it?

jolynch replied:
The messaging rate is so low let's punt on the metric for now? I'm happy to add it as a follow up patch but I'd like to get the pluggability and probes in first.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+     * probeTimerMillis members set.
+     */
+    @VisibleForTesting
+    static void calculateProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples, long intervalMillis) {
```
Paranoia would be to calculate the actual elapsed interval.

jolynch replied:
Yea if we end up moving the ping scheduling to the ping threadpool this will basically be mandatory imo.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+        // Split calculation of probe timers from sending probes for testability
+        calculateProbes(samples, dynamicLatencyProbeInterval);
+
+        if (!StorageService.instance.isGossipActive())
```
Same question here, should this be at the beginning as well?

jolynch replied:
This needs to be here so that the bookeeping clocks keep ticking. Basically we do the calculation but don't send any probes (respecting a disabled gossip indicating we shouldn't be sending messages).

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+        if (updateScoresScheduler != null)
+            updateScoresScheduler.cancel(false);
+        if (updateSamplesScheduler != null)
+            updateSamplesScheduler.cancel(false);
```
Probably don't want to race with this so you should wait until it's completed using get()?

jolynch replied:
Ok, I added a cancel+wait method to do this, and added in some synchronization around `close`/`open` that I believe is is needed to ensure that this method works 100% of the time during snitch updates.

This actually fixes a bug I believe exists in trunk where we just hoped that we didn't get an exception on Snitch construction in https://github.com/apache/cassandra/blob/f3609995c09570d523527d9bd0fd69c2bc65d986/src/java/org/apache/cassandra/service/StorageService.java#L4975-L4986

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+                samples.remove(entry.getKey());
+            }
+
+            if (delay > 0 && millisSinceLastRequest < MAX_PROBE_INTERVAL_MS &&
```
Can remove the millis since last request check if you remove it in calculateProbes

jolynch replied:
Yea I don't think this check is needed anymore since I separated out the two pieces, removed.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
+        public AtomicLong millisSinceLastRequest = new AtomicLong(MAX_PROBE_INTERVAL_MS);
+        public AtomicLong millisSinceLastMeasure = new AtomicLong(0);
+        public long probeTimerMillis = 0;
+        public ScheduledFuture<?> probeFuture = null;
```
Technically should be volatile for toString()

jolynch replied:
Done.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+
+            if (delay > 0 && millisSinceLastRequest < MAX_PROBE_INTERVAL_MS &&
+                (measurement.probeFuture == null || measurement.probeFuture.isDone()) &&
+                !updateSamplesScheduler.isCancelled())
```
The cancelled check can go? It's already pretty racy and redundant.

jolynch replied:
Ok, I thought that it might be good to check just to be safe but sure I took it out.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        schedulePings(samples);
```
Can we stick to just probes? So name it scheduleProbes?

jolynch replied:
Sure, done.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
+        // be ranked at all and therefore we won't probe them.
+        public AtomicLong millisSinceLastRequest = new AtomicLong(MAX_PROBE_INTERVAL_MS);
+        public AtomicLong millisSinceLastMeasure = new AtomicLong(0);
+        public long probeTimerMillis = 0;
```
This would make a lot more sense to me as nextProbeDelay? Since we are tracking how far into the future to delay the next probe?

jolynch replied:
Done.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+            {
+                if (measurement.probeTimerMillis == 0)
+                {
+                    measurement.probeTimerMillis = intervalMillis;
```
assert probe future is null here?

jolynch replied:
It might not be null if we had scheduled a 10 minute probe and then got a real reading.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+                }
+                else if (measurement.probeFuture != null && measurement.probeFuture.isDone())
+                {
+                    measurement.probeTimerMillis = Math.min(MAX_PROBE_INTERVAL_MS, measurement.probeTimerMillis * 2);
```
So this is going to only schedule the next probe MIN_PROBE_INTERVAL_MS after the previous probe has been sent? So every probe is going to have MIN_PROBE_INTERVAL_MS delay added to it on top of whatever the delay is supposed to be?

Seems like the first probe always goes out after 2x MIN_PROBE_INTERVAL_MS? It takes MIN_PROBE_INTERVAL_MS before we check if it is needed and then it's delayed for 2x MIN_PROBE_INTERVAL_MS when the right thing to do is probe immediately.

aweisberg replied:
I take this back. I think being a bounded amount slower at scheduling pings is fine and I like how things are relatively simple right now.

jolynch replied:
Ack, +1

aweisberg reviewed (changes requested):
This is pretty close. I think the DynamicSnitch approach to sending probes is sound although I have some concerns about whether it can schedule probes at the configured rate since it's checking pretty infrequently if it needs to send them. Maybe check every minimum interval / 2 just to keep it simple? 

I looked into having the probes tasks reschedule themselves and I think it's pretty thorny compared to having a thread come along every once in a while and calculate what to do. A whole bunch of CAS loops implementing cancellation and cancellation/replacing the future and that's a bit scary because you can end up scheduling the multiple probes for the same the endpoint.

I'm finishing up reviewing the tests now. Can you add unit tests for reconfiguration of the via the StorageService update method? Maybe make sure that the snitch records correctly after, that there are no orphaned tasks in the STPE, that JMX registration works?

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -23,128 +23,300 @@
+    public static final long MIN_PROBE_INTERVAL_MS = Long.getLong("cassandra.dynamic_snitch_min_probe_interval_ms", 60 * 1000L) ;
+    // The probe rate is set later when configuration is read
+    protected static final RateLimiter probeRateLimiter = RateLimiter.create(1);
+    protected static final DebuggableScheduledThreadPoolExecutor latencyProbeExecutor = new DebuggableScheduledThreadPoolExecutor(1, "LatencyProbes", Thread.MIN_PRIORITY);
```
I am kind of shocked at the number of places we use thread priorities. I have to wonder if anyone has measured the impact of doing that.

jolynch replied:
I took it out per your recommendation.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -60,57 +130,229 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
+        EndpointsForRange orderBefore = full(hosts[2], hosts[1], hosts[3], hosts[4]);
+
+        Map<InetAddressAndPort, DynamicEndpointSnitch.AnnotatedMeasurement> measurements = dsnitch.getMeasurementsWithPort();
+        // At this point we haven't ranked any of hosts so we should need no probes even
```
even?

jolynch replied:
re-worded to make it clearer.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -60,57 +130,229 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
+        {
+            dsnitch.receiveTiming(hosts[1], 20, LatencyMeasurementType.READ);
+            dsnitch.receiveTiming(hosts[2], 10, LatencyMeasurementType.READ);
+            Util.assertRCEquals(orderBefore, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3], hosts[4])));
```
Why would the sort order change if we haven't updated the scores?

jolynch replied:
It shouldn't change, I'm mostly just calling that for the `sortedByProximity` call (might as well check the return value while we're calling it).

I added in a call to `dsnitch.updateScores()` after the `receiveTimings` just to double triple check.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -60,57 +130,229 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
+        }
+        dsnitch.updateScores();
+
+        int probesBefore = (int) ((DynamicEndpointSnitch.MIN_PROBE_INTERVAL_MS / dsnitch.dynamicLatencyProbeInterval) + 1);
```
Probes before what? (Rename?)

jolynch replied:
Ack, renamed to `intervalsBeforeProbesStart` to try to make this clearer.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -60,57 +130,229 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
+            measurements.get(hosts[4]).probeFuture.cancel(false);
+        }
+
+        // Since we haven't measured, we should stop probing
```
Haven't requested right? It's true we haven't measured it, but it's not having requested it that makes not probed.

jolynch replied:
Correct, typo on my part. Fixed.

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -729,20 +730,26 @@ void markTimeout(InetAddressAndPort addr)
 
-    public void addLatency(InetAddressAndPort address, long latency)
+    // Used on the local read path
+    public void addLatency(InetAddressAndPort address, long latencyMicros)
```
It's a bit implicit that this is exclusively for adding read latency?

jolynch replied:
Yeah, I think it was technical debt. I just combined them all now and just rely on the ILatencySubscriber to filter out IGNORES.

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -729,20 +730,26 @@ void markTimeout(InetAddressAndPort addr)
+        addLatency(address, latencyMicros, LatencyMeasurementType.READ);
+    }
+
+    private void addLatency(InetAddressAndPort address, long latencyMicros, LatencyMeasurementType usable)
```
Is it usable or is it a type? It's not a boolean anymore.

jolynch replied:
ack, re-named to `latencyMeasurementType`

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+    private static final Logger logger = LoggerFactory.getLogger(DynamicEndpointSnitch.class);
+
+    // Latency measurement and ranking. The samples contain latency measurements and the scores are used for ranking
+    protected boolean registered = false;
```
Should probably have a more specific similar to how mbeanRegistered specifies what is registered.

jolynch replied:
Ok, changed to `registeredForLatencyUpdates`.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        if (!registered)
```
So why do we register in updateScores? A bit weird no?

Also maybe MessagingService should store a list of WeakReferences so we don't need to unregister? Or we should unregister.

jolynch replied:
Good catch, I just copied that over from the old implementation but yeah that's a nasty memory leak when someone calls `updateSnitch` (even before my change if someone changed the DES update frequency live they would leak an instance into the heap and it would keep getting latency updates ...

I'l try to figure out the right way to register for latency updates (I'd also like to be able to register for actual message events as well so we can do least connection load balancing or something reasonable like that in the future, but let's see how much complexity moving this around is first).

jolynch replied:
So after looking into this it boils down to the snitch is loaded at configuration time before StorageService (and implicitly MessagingService), so it shouldn't register at that time. From my testing we were either getting the MessagingService singleton 200ms after the snitch during `updateScores` or we were getting it during StorageService initialization.

I really don't like weak maps in this case, especially since we iterate the subscriber collection frequently. Instead I've refactored so that we use a `CopyOnWriteArraySet` (since we very rarely add or remove latency subscribers but iterate them on basically every request) in `MessagingService` and properly add and remove ourselves from that map. I think this is the cleanest way that maintains the performance requirements but let me know if you disagree.

For now I'm punting on adding anything additional to `ILatencySubscriber`, I think as part of the followup unifying load shedding from the current BackPressure code path and the DES code path makes sense in CASSANDRA-14817

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+
+        // The underlying measurement technique. E.g. a median filter (histogram) or an EMA filter, or ...
+        public final ISnitchMeasurement measurement;
+        public volatile double cachedMeasurement;
```
Does this need to be volatile? Isn't it all one thread?

jolynch replied:
I believe the only potential concurrent access is from the `toString` method if the logger is set to Trace ... doesn't seem likely.

Removing the volatile.

jolynch replied:
Added back in the volatile since it's now accessed from two threads (sendLatencyProbeToPeer and calculateScores)

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+            if (dynamicUpdateInterval != newDynamicUpdateInternal || updateScoresScheduler == null)
             {
-                updateScores();
+                cancelAndWait(updateScoresScheduler, Math.max(1, dynamicUpdateInterval), TimeUnit.MILLISECONDS);
```
When I am doing termination I prefer to block for 365 days. Either you need to block and wait for it to complete (or it's a bug) or you don't in which case you can cancel and move on. In between doesn't make much sense to me.

jolynch replied:
Interesting, I usually opt to either not block or block with a timeout, but I see your point. In this case none of the futures that are being called should block, and it can't block during the initial startup (at constructor time the futures are `null` so they just immediately succeed at cancelling) so I think I'll just block forever although it pains me to do so heh.

I tried just weaving the TimeoutException back up the stack so we could have our cake and eat it too, but there are lots of places that handling it just becomes difficult (e.g. in the backwards compatible mbean interface I'd have to re-throw it as a runtime exception or something which is confusing and not clear that your DES is now in a bad state).

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
-        };
+            catch (CancellationException | InterruptedException | ExecutionException | TimeoutException ignored)
+            {
+                // Exception is expected to happen eventually due to the cancel -> get
```
It seems like only CanellationException should be ignored? All the others shouldn't happen?

jolynch replied:
I believe you are correct, but we also don't care if the latency probes are interrupted or throw and exception (we're just trying to tear down the futures). I'll update the comment accordingly.

aweisberg replied:
So I just realized this exposes the lack of exception handling for these scheduled tasks. They each need top level exception handling. If they don't have it they will stop without warning and the error will be swallowed up by the scheduled future until something calls get(). They should probably have a top level try catch that logs the error and doesn't let it escape so it will continue to retry on schedule.

That also means that all these other exception types again become an unexpected error except for CancellationException. I think it's fine to continue if you get an error, but we should log it.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+        for (AnnotatedMeasurement measurement : samples.values())
         {
-            public void run()
+            cancelAndWait(measurement.probeFuture, PING.getTimeout(), TimeUnit.MILLISECONDS);
```
I don't get why the PING timeout is appropriate here?

Also this should cancel all of them and then wait on them I think.

jolynch replied:
Yea, I suppose that the probe could at most be waiting for the rate limiter interval (in fact it will be since the ratelimiter uses `sleepUninterruptibly` afaik. Per your above comments I removed the timeouts and just wait for it.

Also I reworked the cancellation method to take a list of futures so that we can cancel all of them and then wait on all of them.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+            // only compute the measurement, which probably generates the most garbage (e.g. for this Histogram),
+            // for endpoints that have been recently updated (millisSinceLastRequest) or somewhat recently requested
+            // for ranking.
+            if (annotatedMeasurement.millisSinceLastMeasure.get() < MIN_PROBE_INTERVAL_MS ||
```
Do we need to check millisSinceLast measure here? We could only check since last request. If we have requested the value then we know independently it will be either probed or getting real measurements so it is worth calculating the scores.

I think a comment to that effect and a simpler condition would make this easier to reason about.

jolynch replied:
The problem with just checking `millisSinceLastRequest` here is that we don't calculate any scores at all in startup until something is requested. I can fix this a few ways though, looking into it.

jolynch replied:
Ok, I just ended up doing what you asked and I made the `millisSinceLastRequest` default to zero instead of maxint (so that way we'll calculate scores for all nodes for the first 10 minutes (like we did in the status quo), and then after that we stop calculating.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -21,133 +21,505 @@
+        if (!StorageService.instance.isGossipActive() || !Gossiper.instance.isAlive(to))
+            return;
+
+        probeRateLimiter.acquire(dynamicSampleUpdateInterval);
```
Reminder to fix the bug where this always does one a second.

jolynch replied:
Ack, fixed.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -214,19 +600,27 @@ public String getDatacenter(InetAddressAndPort endpoint)
 
     // Compare endpoints given an immutable snapshot of the scores
-    private int compareEndpoints(InetAddressAndPort target, Replica a1, Replica a2, Map<InetAddressAndPort, Double> scores)
+    public int compareEndpoints(InetAddressAndPort target, Replica a1, Replica a2)
```
This comment seems wrong and the entire scenario is a little confusing?
Why can't this be removed?

jolynch replied:
Not sure, I didn't write it just copied it over. I think the issue is that `IEndpointSnitch` requires `compareEndpoints(InetAddressAndPort, Replica, Replica)` to be defined so that `AbstractEndpointSnitch::sortedByProximity` works properly, but the only time that the DES would call it is if the DES is wrapping another DES (as `compareEndpoints(InetAddressAndPort, Replica, Replica, Map<InetAddressAndPort, Double>)` internally calls the `subsnitch.compareEndpoints` if it can't find a score.

aweisberg replied:
Ah I get it, it's to catch rewrapping I guess. Fine to leave it.

aweisberg commented on src/java/org/apache/cassandra/net/ResponseVerbHandler.java:
```diff
@@ -49,7 +49,7 @@ public void doVerb(MessageIn message, int id)
         else
         {
             //TODO: Should we add latency only in success cases?
```
Do you think we should remove this TODO while are here?

jolynch replied:
What do you think we should record as the latency under a failure case? Timeouts are already handled elsewhere, so we're just left with execution failures. A failure counting as a timeout seems the most reasonable choice (as nodes that are throwing failures would very rapidly be de-preferred), but not sure.

aweisberg replied:
Let's just leave it alone the scope of this is already big enough.

jolynch replied:
+1, ack

aweisberg commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -463,21 +463,33 @@
+     *                                     otherwise just settings are updated. If an empty string is passed then
+     *                                     dynamic snitching is kept with the default implementation.
+     * @param dynamicUpdateInterval        integer, in ms (defaults to the value configured in cassandra.yaml, which defaults to 100)
+     * @param dynamicSampleUpdateInterval  integer, in ms (defaults to the value configured in cassandra.yaml, which defaults to 1,000)
```
Documenting defaults in too many locations means the docs are likely to be obsoleted. Such as when you fix the bug with the probe rate.

jolynch replied:
yeah these defaults aren't even right ... tech debt --

Removing.

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -18,35 +18,99 @@
 package org.apache.cassandra.locator;
 
-import java.io.IOException;
+import java.net.UnknownHostException;
```
It looks like shutting down the DES is covered a little in the settings test, but it would be good if it validated somewhere the futures are cleaned up and there are no scheduled/running tasks in the executors.

aweisberg replied:
My code coverage is a bit wonky today, but is sendPingMessage tested? Is it possible?

jolynch replied:
> but it would be good if it validated somewhere the futures are cleaned up and there are no scheduled/running tasks in the executors.

ack, writing that

> but is sendPingMessage tested? Is it possible?

It's not tested because it'll require the MessagingService, but I can just setup some sinks for testing. Working on it.

jolynch replied:
Ok, I've added tests by using a message sink, but the test code is a bit gnarly (and very "wait until this thread somewhere does a thing"). As currently implemented we do a true "integration" test to make sure that it works e2e but I don't like it very much. If you're ok with it I'd rather just make the latency probe sending method take some functions and then I can just unit test it (like instead of calling messaging service methods directly it would be passed them). Do you prefer the e2e approach or the unit approach?

aweisberg reviewed (changes requested):
This looks great. It really is close. It looks like everything works except for the rate limiter we talked about.

I took a look at the test and coverage and found some minor nits.

jolynch committed:
More fixups

jolynch committed:
Address remaining comments from Ariel

aweisberg commented on src/java/org/apache/cassandra/concurrent/ScheduledExecutors.java:
```diff
@@ -17,43 +17,111 @@
+
+        for (Pair<String, Integer> executorKey : executorsToWaitFor)
+        {
+            if (!executors.get(executorKey).awaitTermination(executorKey.right, TimeUnit.SECONDS))
```
This waits longer then you would expect if multiple executors are hanging.

This also falls into the, well if you didn't need it to shut down why wait for a limited amount of time? If this is for test then it should fail the test if it times out.

Also should this be lumped into the DES changes?

aweisberg commented on src/java/org/apache/cassandra/locator/dynamicsnitch/DynamicEndpointSnitchLegacyHistogram.java:
```diff
@@ -24,6 +24,7 @@
 
 import org.apache.cassandra.locator.IEndpointSnitch;
 import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.net.MessagingService;
```
Unused import

aweisberg commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -1077,7 +1077,6 @@ private void markAlive(final InetAddressAndPort addr, final EndpointState localS
         logger.trace("Sending a EchoMessage to {}", addr);
         IAsyncCallback echoHandler = new IAsyncCallback()
         {
-            @Override
```
Rando removed override?

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -116,8 +118,9 @@
 import org.apache.cassandra.utils.StatusLogger;
 import org.apache.cassandra.utils.UUIDSerializer;
 import org.apache.cassandra.utils.concurrent.SimpleCondition;
+import org.hsqldb.Database;
```
Unused, but also how in the hell did hsql become a dependency? Question not related to this change of course.

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -116,8 +118,9 @@
+import org.hsqldb.Database;
 
-public final class MessagingService implements MessagingServiceMBean
+    public final class MessagingService implements MessagingServiceMBean
```
Rando whitespace

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -538,7 +541,7 @@ public void resetDroppedMessagesMap(String scope)
     private final Map<Verb, DroppedMessages> droppedMessagesMap = new EnumMap<>(Verb.class);
 
-    private final List<ILatencySubscriber> subscribers = new ArrayList<ILatencySubscriber>();
+    private final Set<ILatencySubscriber> latencySubscribers = new CopyOnWriteArraySet<>();
```
Shouldn't this be CoWArrayList for performance since the most common thing is iteration? Not that it matters much.

aweisberg replied:
Also why is this a list when we only ever have one thing being notified? I mean I guess it's fine to catch snitch reconfiguration races, but seems like overkill since we don't really know if that works anyways.

aweisberg commented on src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java:
```diff
@@ -173,34 +173,52 @@ public boolean execute(Set<InetAddressAndPort> peers, Function<InetAddressAndPor
+     * used for internode messaging (that is not gossip).
+     */
+    private void sendPingMessages(Set<InetAddressAndPort> peers, LatencyMeasurementType latencyMeasurementTypeForSnitch,
+                                  Function<MessageIn, Void> response)
```
Consumer is better for this. No need to use Void.

aweisberg commented on src/java/org/apache/cassandra/concurrent/ScheduledExecutors.java:
```diff
@@ -17,43 +17,111 @@
+        {
+            if (!executors.get(executorKey).awaitTermination(executorKey.right, TimeUnit.SECONDS))
+            {
+                logger.warn("Failed to wait for {} to shutdown", executorKey.left);
```
Definitely an error. Something was supposed to happen and it didn't.

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -5110,6 +5113,11 @@ else if (epSnitchClassName == null || epSnitchClassName.isEmpty())
         updateTopology();
     }
 
+    public void doLocalReadTest()
```
Unused?

aweisberg commented on test/unit/org/apache/cassandra/service/StorageServiceSettingsTest.java:
```diff
@@ -43,6 +48,8 @@
     private static final int UPDATE_INTERVAL = 103;
     private static final int UPDATE_SAMPLE_INTERVAL = 2002;
     private static final double BADNESS = 0.23;
+    private static final String TP_NAME = "LatencyProbes";
```
Unused?

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -262,8 +298,13 @@ public void testLatencyProbeNeeded()
         assertTrue(measurements.containsKey(hosts[3]));
-        Gossiper.instance.markDead(hosts[3], Gossiper.instance.getEndpointStateForEndpoint(hosts[3]));
+        // mark the node as removed
+        Gossiper.instance.injectApplicationState(hosts[3], ApplicationState.STATUS_WITH_PORT,
```
Just curious why this had to change?

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -458,17 +490,38 @@ void scheduleProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples)
-        probeRateLimiter.acquire(dynamicSampleUpdateInterval);
+        probeRateLimiter.acquire();
 
+        // We don't have a good way of estimating the localhost latency right now, so instead we give it
```
The latency probes are just a measure of distance, so in the absence of measurements we measure the  distance and occasionally send a request based on distance to see what the actual performance is? Seems like we could just measure the time to roundtrip a task through the read stage and call it good?

Other non-distance measurements don't seem any more valid to me.

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -687,8 +731,16 @@ public String getSubsnitchClassName()
     public List<Double> dumpTimings(String hostname) throws UnknownHostException
+    {
+        List<Double> micros = dumpTimingsMicros(hostname);
+        return micros.stream().map(s -> s / 1000.0).collect(Collectors.toList());
```
You could use TimeUnit here and in some of the config values at the top of the class.

aweisberg commented on src/java/org/apache/cassandra/locator/dynamicsnitch/DynamicEndpointSnitchLegacyHistogram.java:
```diff
@@ -0,0 +1,73 @@
+            long millisSinceLastMeasure = measurement.millisSinceLastMeasure.getAndAdd(dynamicSampleUpdateInterval);
+            if (millisSinceLastMeasure >= MAX_PROBE_INTERVAL_MS)
+            {
+                samples.clear();
```
I am a bit confused as to why we are clearing samples if we have latency probes? I thought latency probes replace the sample clearing behavior?

jolynch replied:
This is the legacy implementation in case folks want the old behavior back.

aweisberg replied:
Sure but I thought this was the part we wanted to drop by adding probing. We don't need to clear the samples the probes should cause it to correct over time?

aweisberg commented on test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java:
```diff
@@ -60,57 +160,480 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
+        Util.assertRCEquals(order, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3])));
+
+        // make hosts[2] as bad as hosts[1]
+        setScores(dsnitch, 4, allHosts, 15, 20, 10);
```
Why did the number of rounds change in each of these test cases?

aweisberg commented on src/java/org/apache/cassandra/concurrent/ScheduledExecutors.java:
```diff
@@ -17,43 +17,111 @@
-            executor.shutdownNow();
-        for (ExecutorService executor : executors)
-            executor.awaitTermination(60, TimeUnit.SECONDS);
+        List<Pair<String, Integer>> executorsToWaitFor = new ArrayList<>();
```
Is this going to leave behind the shut down executor preventing it from being recreated in the same process?

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -214,30 +659,22 @@ public String getDatacenter(InetAddressAndPort endpoint)
+        // If we don't have latency information about one or more of the replicas, trust the subsnitch until we have
+        // latency information.
+        if (scored1 == null || scored2 == null || scored1.equals(scored2))
             return subsnitch.compareEndpoints(target, a1, a2);
```
I am wondering if there are cases where this breaks. Like if the subsnitch returns a different answer depending on the order of the arguments. Guess we can deal with that if it happens.

aweisberg reviewed (changes requested):
Mostly minor changes requested.
